### PR TITLE
Fix key modifier handling in Web backends [backport 1.4.x]

### PR DESCRIPTION
**Pull request by blink1073**
Addresses #4028.
Fixes an earlier mishandling of the `key_release` event.
Handles all of the standard keys, including `shift` modifiers.
Also properly handles `shift+enter` event in `nbagg`, moving to the next cell.

Test script:

```python
import matplotlib
matplotlib.use('webagg')
import numpy as np
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot(111)
ax.plot(np.random.rand(10))


def onkeyevent(event):
    fig.suptitle('%s: %s' % (event.name, event.key))
    fig.canvas.draw()

cid = fig.canvas.mpl_connect('key_press_event', onkeyevent)
cid = fig.canvas.mpl_connect('key_release_event', onkeyevent)
plt.show()
```

### CODE REVIEW

File: lib/matplotlib/backends/web_backend/mpl.js [hunk 1]
```diff
@@ -459,20 +459,30 @@ mpl.figure.prototype.mouse_event = function(event, name) {
...
+        value += "control";
+    }
+    else if (event.keyCode == 18) {
+        value += "alt";
+    }
+    else if (event.keyCode == 16) {
+        value += "shift";
+    }
+    else {
+        value += String.fromCharCode(event.keyCode).toLowerCase();
```
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: This does raise the question why we don't pass 'shift+e' and as 'E', but I suspect that there is a good reason for this....(goes to look how qt does it).

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: AFAIK, Qt just feeds you a key code, and leaves it up to you to capitalize it based on the shift modifier.

**Comment by tacaswell** [comment]
Great! That seems to have fixed it.  The only dicey thing is that it seems to generate tons of events if you hold a key down (despite 'keydown' claiming to only fire once).

attn @pelson @mdboom

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 1 above)
Comment (pelson) [inline lib/matplotlib/backends/web_backend/mpl.js]: Is this behaviour consistent with each of the QtAgg, TkAgg and GtkAgg backends? (with regards to shift?)

Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: The qt backend applies the shift, have not checked the others.

Comment (WeatherGod) [inline lib/matplotlib/backends/web_backend/mpl.js]: I think all backends apply the shift. I have keymaps that have capital
letters.

**Comment by blink1073** [comment]
@tacaswell, I added a guard against multiple key press events.

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 1 above)
Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: So we want to remove the `.toLowerCase()` then?

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: Well, we'd also need to remove "shift" from the `value`.

Comment (WeatherGod) [inline lib/matplotlib/backends/web_backend/mpl.js]: I think only if the character can be capitalized.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: I'll play around with it.

**Comment by blink1073** [comment]
We can now handle `shift` state and tell `a` vs `A`.  However, the Javascript [key codes](http://www.cambiaresearch.com/articles/15/javascript-char-codes-key-codes) do not map to ASCII codes.  We'd have to use a LUT to handle keys like `Home` and the function keys.

**Comment by blink1073** [comment]
I've got most of the key identifiers working in both Firefox and Chrome.  The only one that didn't work was '~' in Chrome.

**Comment by blink1073** [comment]
I believe I got all key combinations right now in both Firefox and Chrome.  I also update the UAT to show the event type and metadata in the figure title.  
I tried booting into Windows to try IE, but ran into several problems, then Windows told me to restart, and I rage quit...

**Comment by blink1073** [comment]
Also tested in IE 11.

**Comment by blink1073** [comment]
jQuery provides the `which` attribute, so I used that consistently and pushed the key handling logic into Python.  I used a LUT for those values that could not be easily teased out.

**Comment by blink1073** [comment]
@tonysyu, would you mind testing this with Safari using the [UAT](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backends/web_backend/nbagg_uat.ipynb) notebook?

**Comment by tonysyu** [comment]
@blink1073: This seems to behave the same on Safari and Chrome.

One small issue (independent of browser) is that this branch changes the behavior where you can simply advance through the notebook using Shift+Enter: When you hit a cell with a figure, Shift+Enter executes the cell but doesn't advance to the next cell. This might be unavoidable, but I don't know enough about what's going on underneath the hood.

**Comment by tacaswell** [comment]
That is probably because nbagg grabs focus on plot (to match the behavior of the GUI backends) and when it get focus it disables all of IPython's key event handling (so that _we_ can get the key events and pass them down to the python layer).

Maybe we want to special case `'shift-enter'` and bounce that back up to IPython?

**Comment by blink1073** [comment]
@tacaswell, seems a bit magical to me, but I don't feel that strongly.

**Comment by tacaswell** [comment]
From watching people use notebooks being able to 'shift-enter' your way down a page _seems_ to be a major use-case which I would rather not break.

cc @danielballan @rgbkrk am I reading the importance of this use-case right?

**Comment by blink1073** [comment]
I'm looking into it now, but I'm a bit stuck, see 0efea3b.
Passing the event to the IPython keyboard manager just refreshes the existing output.

**Comment by blink1073** [comment]
It works now, spoofing a "J" command, which tells IPython to go to the next cell.

**Comment by rgbkrk** [comment]
Yeah, `shift-enter` is the typical flow for a lot of people (including myself) for running through a notebook.

**Comment by danielballan** [comment]
Yes that is an important use case.

Under the hood, cell advancement is related to the mode (edit or command). That is vague; I'm willing to take a closer look if needed.

**Comment by blink1073** [comment]
Thanks @danielballan, I added an explicit call to `command_mode()` prior to `handle_keydown()`

**Comment by blink1073** [comment]
@tacaswell, I think I'm done monkeying around :smile:   Rebased.

**Comment by danielballan** [comment]
@blink1073 Great.

File: lib/matplotlib/backends/web_backend/mpl.js [hunk 2]
```diff
@@ -458,21 +458,35 @@ mpl.figure.prototype.mouse_event = function(event, name) {
     return false;
 }
 
+mpl.figure.prototype._key_event_extra = function(event, name) {
```
Comment (pelson) [inline lib/matplotlib/backends/web_backend/mpl.js]: Do we need to start doing JS "docstrings"?

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: Better?

**Comment by tacaswell** [comment]
@blink1073 have a PR against your branch, there is a missing `manager =` assignment for 'old' IPython versions.

**Comment by tacaswell** [comment]
backported as 1fa015318cfbc073c1f10f3cc836ce1e6ecb612d

**Pull request merged**